File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/errors/Pattern.scala
package org.enso.compiler.core.ir
package expression
package errors

import org.enso.compiler.core.{IR, Identifier}

import java.util.UUID

/** A representation of an error resulting from wrong pattern matches.
  *
  * @param originalPattern pattern that resulted in the error
  * @param reason          the cause of this error
  * @param passData        the pass metadata associated with this node
  * @param diagnostics     compiler diagnostics for this node
  * @return a copy of `this`, updated with the specified values
  */
sealed case class Pattern(
  originalPattern: org.enso.compiler.core.ir.Pattern,
  reason: Pattern.Reason,
  passData: MetadataStorage      = new MetadataStorage(),
  diagnostics: DiagnosticStorage = DiagnosticStorage()
) extends Error
    with Diagnostic.Kind.Interactive
    with org.enso.compiler.core.ir.Pattern
    with LazyId {
  override def mapExpressions(
    fn: java.util.function.Function[Expression, Expression]
  ): Pattern =
    copy(originalPattern = originalPattern.mapExpressions(fn))

  override def setLocation(location: Option[IdentifiedLocation]): Pattern =
    copy(originalPattern = originalPattern.setLocation(location))

  /** Creates a copy of `this`.
    *
    * @param originalPattern the pattern that resulted in the error
    * @param reason          the cause of this error
    * @param passData        the pass metadata associated with this node
    * @param diagnostics     compiler diagnostics for this node
    * @param id              the identifier for the new node
    * @return a copy of `this`, updated with the specified values
    */
  def copy(
    originalPattern: org.enso.compiler.core.ir.Pattern = originalPattern,
    reason: Pattern.Reason                             = reason,
    passData: MetadataStorage                          = passData,
    diagnostics: DiagnosticStorage                     = diagnostics,
    id: UUID @Identifier                               = id
  ): Pattern = {
    val res = Pattern(originalPattern, reason, passData, diagnostics)
    res.id = id
    res
  }

  /** @inheritdoc */
  override def duplicate(
    keepLocations: Boolean   = true,
    keepMetadata: Boolean    = true,
    keepDiagnostics: Boolean = true,
    keepIdentifiers: Boolean = false
  ): Pattern =
    copy(
      originalPattern = originalPattern
        .duplicate(
          keepLocations,
          keepMetadata,
          keepDiagnostics,
          keepIdentifiers
        ),
      passData =
        if (keepMetadata) passData.duplicate else new MetadataStorage(),
      diagnostics =
        if (keepDiagnostics) diagnostics.copy else DiagnosticStorage(),
      id = if (keepIdentifiers) id else null
    )

  override def message(source: (IdentifiedLocation => String)): String =
    reason.explain

  override def diagnosticKeys(): Array[Any] = Array(reason)

  override val location: Option[IdentifiedLocation] =
    originalPattern.location

  override def children: List[IR] = List(originalPattern)

  override def showCode(indent: Int): String =
    originalPattern.showCode(indent)
}

object Pattern {

  /** A representation of the reason the pattern is erroneous.
    */
  sealed trait Reason {

    /** Provides a human-readable explanation of the error.
      *
      * @return
      */
    def explain: String
  }

  /** A reason for pattern failing due to wrong arity.
    *
    * @param consName the constructor name.
    * @param expected expected field count.
    * @param actual   actual field count.
    */
  case class WrongArity(consName: String, expected: Int, actual: Int)
      extends Reason {
    override def explain: String =
      s"Wrong number of fields when matching on $consName." +
      s" Expected $expected fields, but provided $actual"
  }
}

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/expression/errors/Pattern.java
package org.enso.compiler.core.ir.expression.errors;

import org.enso.compiler.core.ir.Diagnostic$Kind$Interactive;
import org.enso.compiler.core.ir.DiagnosticStorage;
import org.enso.compiler.core.ir.IRKind;
import org.enso.compiler.core.ir.IdentifiedLocation;
import org.enso.compiler.core.ir.MetadataStorage;
import org.enso.compiler.core.ir.expression.Error;
import org.enso.runtime.parser.dsl.GenerateFields;
import org.enso.runtime.parser.dsl.GenerateIR;
import org.enso.runtime.parser.dsl.IRChild;
import org.enso.runtime.parser.dsl.IRField;
import scala.Function1;
import scala.Option;

/** A representation of an error resulting from wrong pattern matches. */
@GenerateIR(
    interfaces = {
      Error.class,
      Diagnostic$Kind$Interactive.class,
      IRKind.Primitive.class,
      org.enso.compiler.core.ir.Pattern.class
    })
public final class Pattern extends ErrorPatternGen {

  /**
   * @param originalPattern pattern that resulted in the error
   * @param reason the cause of this error
   */
  @GenerateFields
  public Pattern(
      @IRChild org.enso.compiler.core.ir.Pattern originalPattern,
      @IRField Reason reason,
      MetadataStorage passData,
      DiagnosticStorage diagnostics) {
    super(originalPattern, reason, passData, diagnostics);
  }

  public static Pattern create(org.enso.compiler.core.ir.Pattern originalPattern, Reason reason) {
    return builder().originalPattern(originalPattern).reason(reason).build();
  }

  @Override
  public String message(Function1<IdentifiedLocation, String> source) {
    return reason().explain();
  }

  @Override
  public Object[] diagnosticKeys() {
    return new Object[] {reason()};
  }

  @Override
  public IdentifiedLocation identifiedLocation() {
    return originalPattern().identifiedLocation();
  }

  @Override
  public Pattern setLocation(Option<IdentifiedLocation> location) {
    var newOrigPattern = originalPattern().setLocation(location);
    return builder(this).originalPattern(newOrigPattern).build();
  }

  @Override
  public String showCode(int indent) {
    return originalPattern().showCode(indent);
  }

  public sealed interface Reason {
    String explain();
  }

  /**
   * A reason for pattern failing due to wrong arity.
   *
   * @param consName the constructor name.
   * @param expected expected field count.
   * @param actual actual field count.
   */
  public record WrongArity(String consName, int expected, int actual) implements Reason {

    @Override
    public String explain() {
      return "Wrong number of fields when matching on "
          + consName
          + ". Expected "
          + expected
          + " fields, but provided "
          + actual;
    }
  }

  /**
   * An error when a project-private constructor is used in the pattern.
   *
   * @param consName Name of the constructor. Does not have to be fully qualified.
   * @param callerProject The project name of the caller.
   * @param calleeProject The project name of the callee. The constructor is in this project.
   */
  public record PrivateConstructor(String consName, String callerProject, String calleeProject)
      implements Reason {

    @Override
    public String explain() {
      return "Project-private constructor '"
          + consName
          + "' in project '"
          + calleeProject
          + "' cannot be used from project '"
          + callerProject
          + "'";
    }
  }
}
